Update config sources manager (fixes/changes) (#3687)
open-telemetry/opentelemetry-collector: config/internal/configsource/README.md

| +-------------------------------------------------+ |                |                                 |+   
+-----------------------------------------------------+                +---------------------------------+    
                       |                                                                |                     
                       |                                                                |                     
                       +----------------------------------------------------------------+                     
                                                     |                                                        
                                                     |                                    +------------------+
                                                     v                                   +------------------+|
                                     +-------------------------------+                  +------------------+||
                                     |                               |                  |                  |||

>                                      |      configsource.Manager     |                  | configsource Obj ||+

Reviewer: What's the relations between `configsource.Manager` and `configsource Obj` here? There is no arrow, os it is a bit unclear what happens here. Perhaps explain in the text?

Author: Trying to express that `configsource.Manager` creates the `configsource Obj` from the initial configuration. I will add some text and perhaps a separate diagram.

Author: I opted to remove it from this diagram and explain with text below.

                          | |      exporters: [otlp]                          | |                             
                          | +-------------------------------------------------+ |                             
                          +-----------------------------------------------------+                             
                                                                                                              
                                                                                                                                                                                                                            
```

## Watching for Updates
After the configuration was processed the `configsource.Manager` can be used as a single point to watch for updates in
the configuration data retrieved via the config sources used to process the “initial” configuration and to generate
the“effective” one.

The `configsource.Manager` does that by wrapping calls to the `WatchForUpdate()` method of each `configsource.Retreived`
object that was used during the configuration processing. It also controls the lifecycle of all `configsource.Session`

> objects created to get the `configsource.Retrieved` objects. 

Reviewer: Maybe expand a bit more on what is `configsource.Retreived` and what is `configsource.Session`. Neither of these are visible on the diagram. Is it possible to add them or have a different diagram that shows them?

Author: Will do.

Author: Done. Remember that there is also some info about [configsorce.Retrieved](https://github.com/open-telemetry/opentelemetry-collector/blob/b336d3df8b9d483c1e07e4bc216d788289da7e9d/config/experimental/configsource/component.go#L83) and [configsource.Session](https://github.com/open-telemetry/opentelemetry-collector/blob/b336d3df8b9d483c1e07e4bc216d788289da7e9d/config/experimental/configsource/component.go#L57) at the `config/experimental/configsource` package. We should probably re-organize these packages, but after we get the manager code updated.

# Config Source Manager
The `configsource.Manager` handles the use of multiple [config source objects](../../experimental/configsource/component.go)
simplifying configuration parsing,
monitoring for updates, and the overall life-cycle of the used config sources.

## Usage Overview
The public API of the `configsource.Manager` is described in `godoc` comments at the
[`manager.go`](./manager.go) file.
Below is an overview of its usage, which consists of two separate phases:

1. Configuration Processing
2. Watching for Updates

## Configuration Processing

> The `configsource.Manager` receives as input a set of config source factories and a `configparser.Parser` that

Reviewer: ```suggestion
The `configsource.Manager` receives as input a set of config source factories and a local `configparser.ConfigMap` that
```

> will be used to generate the resulting, or effective, configuration also in the form of a `configparser.Parser`,

Reviewer: ```suggestion
will be used to generate the resulting, or effective, configuration also in the form of a `configparser.ConfigMap`,
```

that can be used by code that is oblivious to the usage of config sources.

```terminal
+-----------------------------------------------------+                                                       

> |              configparser.Parser                    |                                                       

Reviewer: ```suggestion
|              configparser.ConfigMap                    |                                                       
```

+---------+         +--------------+         +---------+          +-----------+
    |                      |                      |                     |      
    |      NewSession      |                      |                     |      
    |--------------------->+--+                   |                     |      
    |                      |  |                   |                     |      
    |                      |  |------------------>+--+                  |      
    |                      |  |                   |  |                  |      
    |                      |  |                   |  |                  |      
    |                      |  |<------------------+--+                  |      
    |                      |  |                   |                     |      
    |<---------------------+--+                   |                     |      
    |  "Session object"    |                      |                     |      
    |                      |                      |                     |      
    |                      |       Retrieve       |                     |      
    |-------------------------------------------->+--+                  |      

>     |                      |                      |  |   NewRetrieved   |      

Reviewer: `NewRetrieved` seems to be only used for tests, so unclear why it is in this diagram, which I assume specifically is about not tests.

Author: This becomes a mute point with some other refactors on the repo: the Retrieved type here was just a helper for common functionality. Now the same is expressed as interfaces, `NewRetrieved` will disappear.
